enhancements/installer/assisted-installer-in-cluster.md

```diff
@@ -0,0 +1,641 @@
+obtain the discovery live ISO and use my own methods to boot it, or I can use
+the baremetal-operator to boot the live ISO automatically.
+
+#### Multi-cluster
+
+As a user of Red Hat's multi-cluster mangement tools, I can use the assisted
+installation agent-based workflow to create clusters from a pool of bare metal
+inventory.
+
+### Implementation Details/Notes/Constraints
+
+#### Assisted Installer k8s-native API
+
+The Assisted Installer has a [REST
+API](https://generator.swagger.io/?url=https://raw.githubusercontent.com/openshift/assisted-service/master/swagger.yaml)
+that is available at cloud.redhat.com. The service is implemented with a
+traditional relational database. In order to integrate with OpenShift
+infrastructure management tooling, it is necessary to additionally expose its
+capabilities as Kubernetes-native APIs. They will be implemented as Custom
+Resource Definitions and a controller with the assisted service's capabilities.
+
+For the first deliverable, a local deployment of the backing service will
+include new controllers that will operate to achieve the desired state as
+declared in CRDs. Thus the service will expose both REST and Kubernetes-native
+APIs, and both API layers will call the same set of internal methods. The local
+service will include a database; sqlite if possible, or else postgresql.
+
+    --------------------------
+    |  REST API  |  k8s API  |
+    --------------------------
+    | Assisted Service logic |
+    --------------------------
+        |               |
+        v               v
+      SQL DB        File System
+                      storage
+
+For modest clusters that do not have persistent storage available but need
+day-2 ability to add nodes, the database will need to be reconstructable in
+case its current state is lost. Thus the CRs will be the source of truth for
+the specs, and the controller will upon statup first ensure that the DB
+reflects the desired state in CRs. The source of truth for the status, however,
+is the actual state of the agents running on the hosts being installed and not
+necessarily what was previously recorded.
+
+Agents that are running on hardware and waiting for installation-related
+instructions will continue to communicate with the assisted service using the
+existing REST API. In the future that may transition to using a CRD.
+
+**AgentDrivenInstallationImage**
+This resource represents a discovery image that should be used for booting
+hosts. In the REST API this resource is embedded in the "cluster" resource, but
+for kubernetes-native is more natural for it to be separate.
+
+Spec:
+ * clusterName: string
+ * sshPublicKey: string
+ * pullSecretName: string (reference to k8s secret)
+ * ignitionOverrides: object
+ * staticIpConfiguration: object
+
+Status:
+ * state: string
+ * sizeBytes: integer
+ * downloadUrl: string
+ * createdAt: string
+ * expiresAt: string
+
+**AgentDrivenClusterInstallation**
```

Comment:
One point of clarification that would be useful to make here - that this resource will play no part in the standalone cluster use cases, that it will only play a role in the multi-cluster use cases

Reply:
And then in the context of multi-cluster installs, we should clearly distinguish between the purpose of ClusterDeployment and AgentDrivenClusterInstallation - right now, they appear to overlap in difficult to justify ways

The "create cluster" workflow below highlights this when it says:

```
2. User creates AssistedInstallCluster resource describing their desired cluster.
3. User creates ClusterDeployment describing a new cluster.
```

What is the user describing about the cluster with AgentDrivenClusterInstallation that should not be described by ClusterDeployment?

Reply:
Another example would be e.g. machineNetworkCidr - if you created a ClusterDeployment referencing an install-config with a particular machine network (or with `hiveutil create-cluster --machine-network`), and separately AssistedInstallCluster listed a different machine network, which would win?

Reply:
We're looking at having at least 3 clusters in play, ACM, Hive, Assisted (and in a parallel stream a Hypershift cluster). There will likely be substantial if not complete duplication between them, as a Hive caller should be able to fully specify their cluster in Hive. They should not need to create CRs for the Assisted operator directly. (as far as I am aware)

We talked about doing this so teams can develop independently and as quickly as possible, and components remain useful in isolation. I'm still a bit nervous about this and not sure if it's the ideal approach. (and we should always strive to reduce the load on etcd, so I don't love duplicating all of this everywhere)

Reply:
> They should not need to create CRs for the Assisted operator directly. (as far as I am aware)
Discussions today have revealed this to be less of a mandate than we may have thought. Let's continue to discuss, and I think we can agree on something reasonable. Making duplicate pass-through APIs in hive to mirror the complete APIs of agent-based installation probably doesn't make sense.

Reply:
After a chat with @eparis and @derekwaynecarr yesterday I think we should discuss eliminating the middle API and operator, if not now, then after prototyping. The end resulting topology should be the same, and it feels like the only reason we're duplicating those APIs is to not have to figure out how to work together. From Hive POV it doesn't feel like it makes sense that we would wrap a REST API with Kube APIs, and then wrap those Kube APIs with more Kube APIs. I don't think we have a need for a standalone Assisted operator, and I think that would be a misunderstanding of what exactly Hive is. It is quite small and lightweight, easy to get running, and it's job is to provision configurations of OpenShift at scale.

I would pitch that once we've got a prototype up (or sooner if desired), we start pushing that code to hive. Review process will be difficult for both sides but there's a big benefit there in that we can help translate that imperative API to a declarative Kube API as my team has been at this for a few years. 

In this scenario I would assume that the current assisted install code base remains separate, as this allows the really deep domain logic for performing these installs to stay in the teams with that knowledge. Presumably this would also still need an OperatorHub operator to get the REST API up on a hub cluster, just with a much more limited scope.